rfcs/20230515-DynamicEmbedding.md

```diff
@@ -0,0 +1,251 @@
+front - TorchRec with a native embedding distribution and a dynamic
+embedding solution & HugeCTR (Merlin) with a highly-performant
+embedding caching strategy. This makes it essential for TensorFlow to
+introduce a native solution to stay competitive in the market.
+
+## Design Proposal
+In this design approach, the DynamicEmbedding layer is composed of two
+layers: the DynamicLookup layer and the Embedding layer. The
+DynamicLookup layer is responsible for the following tasks:
+    * Maintaining a vocabulary table using an eviction policy that is 
+    updated based on input pattern.
+    * Performing vocabulary lookup for the given input and returning
+    integer indexes.
+    * The index is then passed to the Embedding layer, which looks
+    up the embedding vector. The Embedding layer is responsible for
+    the following tasks:
+        + Looking up the embedding vector for the given integer index.
+        + Returning the embedding vector.
+The embedding vector is then used by the subsequent layer in the
+neural network. The Dynamic Embedding layer is used in conjunction
+with UpdateEmbeddingCallback. The callback is triggered at a
+predetermined time interval. It aggregates the Dynamic vocabulary
+table across all workers and updates the vocabulary that is used
+for input lookup across all workers. This ensures that the vocabulary
+is always up-to-date and that all workers are using the same vocabulary. 
+
+
+![DynamicEmbedding](/20230515-DynamicEmbedding/DynamicEmbedding.png)
+
+Here is a deeper look at what is done in DynamicLookup layer and how the
+UpdateEmbeddingCallback updates the embeddings and vocabulary
+The DynamicEmbedding layer identifies and adds unique keys to the dynamic
+vocabulary table for every input passed to it. This table is constantly
+updated based on the eviction policy provided, such as TTL, LFU, or LRU.
+The table is maintained on each worker when used with distributed
+training, and the tables on different workers may be different.
+The UpdateEmbeddingCallback is a timed callback that uses a timed
+thread to create a callback event when the timer expires. The callback
+aggregates the dynamic vocabulary table values across all workers in a
+distributed training setup and updates the vocabulary on all workers. 
+Update the vocab->index mapping(mutable hash table/ tf.Variable) on
+all workers Update/remap the embedding matrix to reflect new
+vocabulary-> index mapping
+    * Old vocab keys will have the same embedding vector 
+    * New vocab keys will have newly initialized embedding vector
+This updated vocabulary is used for lookup in the DynamicLookup layer
+until the callback event is triggered again after the time interval.
+
+![DynamicLookup](0230515-DynamicEmbedding/DynamicLookup.png)
+
+The image below illustrates the workflow when the parameter server
+strategy is used. PSS supports asynchronous training. Each worker
+will have a copy of the vocabulary, which will be consistent across
```

Comment:
If it means each worker will hold a full set of vocabulary that maps the `vocab` to `index`, and the real embedding vectors stored in some PSs with dense format(for example: the `tf.Variable`)?

Reply:
That is correct. Each worker should have a copy of the vocabulary( vocab->index mapping). The embedding variable will be split in distributed servers.

Reply:
Hi @divyashreepathihalli, thank you for your comment!  If we have a full set copy of the `key-index` mapping on each worker, there should be some upper limitations on the `vocabulary size`. To my best knowledge, some `vocabulary size` in some industrial scenarios can be **tens or hundreds of billions**, which causes the memory consumption on GPU/TPU to be significantly large and expensive. The practical method is storing the `key-value` with the format of abstract hashtable in a distributed way like TFRA. Hope it's helpful. Thanks!